website/api/controllers/issue-est-certificate.js

```diff
@@ -0,0 +1,133 @@
+module.exports = {
+
+
+  friendlyName: 'Issue certificate via EST',
+
+
+  description: 'Take a certificate signing request and authentication token, then use the EST protocol to request a certificate to be issued.',
+
+  extendedDescription: '',
+
+
+  inputs: {
+    csrData: {
+      required: true,
+      type: 'string',
+      description: 'CSR'
+    },
+    authToken: {
+      required: true,
+      type: 'string',
+      description: 'Authorization token provided by IdP',
+    }
+  },
+
+
+  exits: {
+
+    success: {
+      description: 'Successfully generated certificate',
+      outputType: {certificate:'string'},
+      outputFriendlyName: 'Certificate',
+    },
+
+    invalidToken: {
+      description: 'The IdP auth token is invalid',
+      statusCode: 403,
+    },
+
+    badRequest: {
+      responseType: 'badRequest'
+    }
+
+  },
+
+  fn: async function ({ csrData, authToken }) {
+    const INTROSPECT_ENDPOINT = sails.config.custom.certIssueIdpIntrospectEndpoint;
+    if (!INTROSPECT_ENDPOINT) {
+      throw new Error('sails.config.custom.certIssueIdpIntrospectEndpoint is required');
+    }
+    const IDP_CLIENT_ID = sails.config.custom.certIssueIdpClientId;
+    if (!IDP_CLIENT_ID) {
+      throw new Error('sails.config.custom.certIssueIdpClientId is required');
+    }
+    const EST_ENDPOINT = sails.config.custom.certIssueEstEndpoint;
+    if (!EST_ENDPOINT) {
+      throw new Error('sails.config.custom.certIssueEstEndpoint is required');
+    }
+    const EST_CLIENT_ID = sails.config.custom.certIssueEstClientId;
+    if (!EST_CLIENT_ID) {
+      throw new Error('sails.config.custom.certIssueEstClientId is required');
+    }
+    const EST_CLIENT_KEY = sails.config.custom.certIssueEstClientKey;
+    if (!EST_CLIENT_KEY) {
+      throw new Error('sails.config.custom.certIssueEstClientKey is required');
+    }
+
+
+    // Ask the IdP to introspect the auth token (ensure it's valid and extract the values).
+    // TODO can this be done with sails.helpers.http? Couldn't figure out how to send form data.
+    const urlencoded = new URLSearchParams();
+    urlencoded.append('client_id', IDP_CLIENT_ID);
+    urlencoded.append('token', authToken);
+    let axios = require('axios');
+    const introspectResponse = await axios({
+      url: INTROSPECT_ENDPOINT,
+      method: 'POST',
+      data: urlencoded,
+    });
```

Comment:
@eashaw please lmk if you know how to use the built-in sails helpers to send form data with an HTTP request. Then I can remove axios.

Reply:
@zwass This should work to send the HTTP request
```suggestion
    const introspectResponse = await sails.helpers.http.sendHttpRequest.with({
      url: INTROSPECT_ENDPOINT,
      method: 'POST',
      enctype: 'application/x-www-form-urlencoded',
      data: {
        client_id: IDP_CLIENT_ID,
        token: authToken
      },
    });
```

Reply:
@eashaw thank you! That was almost it but I also had to change `data` to `body`.